File: Editorial/150_Evaluate_Reverse_Polish_Notation/CSharp/1_Reducing_the_List_In_place.cs
public class Solution {
    private static Dictionary<string, Func<int, int, int>> OPERATIONS =
        new Dictionary<string, Func<int, int, int>>() {
            { "+", (int a, int b) => a + b },
            { "-", (int a, int b) => a - b },
            { "*", (int a, int b) => a * b },
            { "/", (int a, int b) => a / b }
        };

    public int EvalRPN(string[] tokens) {
        int currentPosition = 0;
        while (tokens.Length > 1) {
            while (!OPERATIONS.ContainsKey(tokens[currentPosition])) {
                currentPosition++;
            }

            string operation = tokens[currentPosition];
            int number1 = Int32.Parse(tokens[currentPosition - 2]);
            int number2 = Int32.Parse(tokens[currentPosition - 1]);
            Func<int, int, int> func = OPERATIONS[operation];
            int value = func(number1, number2);
            tokens[currentPosition] = value.ToString();
            List<string> tokenslist = tokens.ToList();
            tokenslist.RemoveRange(currentPosition - 2, 2);
            tokens = tokenslist.ToArray();
            currentPosition--;
        }

        return Int32.Parse(tokens[0]);
    }
}

File: Editorial/0150_Evaluate_Reverse_Polish_Notation/CPP/1_Reducing_the_List_In_place.cpp
class Solution {
private:
    map<string, function<int(int, int)>> OPERATIONS = {
        {"+", [](int a, int b) { return a + b; }},
        {"-", [](int a, int b) { return a - b; }},
        {"*", [](int a, int b) { return a * b; }},
        {"/", [](int a, int b) { return a / b; }}};

public:
    int evalRPN(vector<string>& tokens) {
        int currentPosition = 0;
        while (tokens.size() > 1) {
            while (OPERATIONS.find(tokens[currentPosition]) ==
                   OPERATIONS.end()) {
                currentPosition++;
            }
            string operation = tokens[currentPosition];
            int number1 = stoi(tokens[currentPosition - 2]);
            int number2 = stoi(tokens[currentPosition - 1]);
            auto it = OPERATIONS.find(operation);
            int value = (it->second)(number1, number2);
            tokens[currentPosition] = to_string(value);
            tokens.erase(tokens.begin() + currentPosition - 2,
                         tokens.begin() + currentPosition);
            currentPosition--;
        }
        return stoi(tokens[0]);
    }
};
